detection_rules/misc.py

```diff
@@ -32,6 +55,291 @@
+
+        release_metadata = self._parse_release_metadata()
+        release_metadata.update(assets=assets)
+        release_manifest = ReleaseManifest(**release_metadata)
+
+        return release_manifest
+
+    def _parse_release_metadata(self):
+        """Parse relevant info from GitHub metadata for release manifest."""
+        ignore = ['assets']
+        manual_set_keys = ['author', 'description']
+        keys = [f for f in list(ReleaseManifest.__annotations__) if str(f) not in ignore + manual_set_keys]
+        parsed = {k: self.release.raw_data[k] for k in keys}
+        parsed.update(description=self.release.raw_data['body'], author=self.release.raw_data['author']['login'])
+        return parsed
+
+    def save(self) -> GitReleaseAsset:
+        """Save manifest files."""
+        if not self.has_token:
+            raise ValueError('You must provide a token to save a manifest to a GitHub release')
+
+        asset = self.release.upload_asset_from_memory(self.release_manifest_fl,
+                                                      self.manifest_size,
+                                                      self.manifest_name)
+        click.echo(f'Manifest saved as {self.manifest_name} to {self.release.html_url}')
+        return asset
+
+    @classmethod
+    def load(cls, name: str, repo: str = 'elastic/detection-rules', token=None) -> dict:
+        """Load a manifest."""
+        gh_client = GithubClient(token)
+        repo = gh_client.client.get_repo(repo)
+        release = get_gh_release(repo, tag_name=name)
+        asset = next((a for a in release.get_assets() if a.name == f'manifest-{name}.json'), None)
+        return load_json_gh_asset(asset.browser_download_url) if asset else None if asset else None
+
+    @classmethod
+    def load_all(cls, repo: str = 'elastic/detection-rules', include_missing=False, token=None) -> dict:
+        """Load a consolidated manifest."""
+        gh_client = GithubClient(token)
+        repo = gh_client.client.get_repo(repo)
+
+        consolidated = {}
+        for release in repo.get_releases():
+            name = release.tag_name
+            asset = next((a for a in release.get_assets() if a.name == f'manifest-{name}.json'), None)
+            if not asset and not include_missing:
+                continue
+
+            consolidated[name] = load_json_gh_asset(asset.browser_download_url) if asset else None
+
+        return consolidated
+
+    @classmethod
+    def get_existing_asset_hashes(cls, repo: str = 'elastic/detection-rules', token=None) -> dict:
+        """Load all assets with their hashes, by release."""
+        flat = {}
+        consolidated = cls.load_all(repo=repo, token=token)
+        for release, data in consolidated.items():
+            for asset in data['assets'].values():
+                for asset_name, asset_data in asset['entries'].items():
+                    flat.setdefault(release, {})[asset_name] = asset_data['sha256']
```

Comment:
lots of redundant calls, could just separate out the whole setdefault thing

```suggestion
            for asset in data['assets'].values():
                flat_release = flat[release] = {}
                for asset_name, asset_data in asset['entries'].items():
                    flat_release[asset_name] = asset_data['sha256']
```